kgforge/core/archetypes/database.py

```diff
@@ -0,0 +1,145 @@
+    # POLICY Implementations should not add methods but private functions in the file.
+    # POLICY Implementations should pass tests/specializations/databases/test_databases.py.
+
+    def __init__(self, forge : Optional["KnowledgeGraphForge"], source: str, **config) -> None:
+        # POLICY Resolver data access should be lazy, unless it takes less than a second.
+        # POLICY There could be data caching but it should be aware of changes made in the source.
+        self._forge: Optional["KnowledgeGraphForge"] = forge
+        # Model
+        model_config = config.pop("model")
+        if model_config["origin"] == "store":
+            use_values(
+                model_config,
+                config,
+                ["endpoint", "token", "bucket", "vocabulary"],
+            )
+        model_name = model_config.pop("name")
+        model = import_class(model_name, "models")
+        self._model: Model = model(**model_config)
+        self.context = self._model.context()
+        if 'model_context' not in config:
+            config['model_context'] = self._model.context()
+        self.source: str = source
+        self.service: Any = self._initialize_service(self.source, **config)
+
+    def __repr__(self) -> str:
+        return repr_class(self)
+
+    def map(self, resources : Union[List[Union[Resource, str]], Union[Resource, str]],
+            type_ : Optional[Union[str, "DictionaryMapping"]] = None) -> Optional[Union[Resource, List[Resource]]]:
+        mappings = self._model.mappings(self._model.source, False)
+        mapped_resources = []
+        resources = (resources if isinstance(resources, list) else [resources])
+        for resource in resources:
+            if isinstance(resource, Resource):
+               resource_dict = self._forge.as_json(resource) 
+            else:
+                resource_dict = resource
+                resource = self._forge.from_json(resource_dict)
+            if type_ is None:
+                try:
+                    type_ = resource.type
+                except AttributeError:
+                    mapped_resources.append(resource)
+            elif isinstance(type_, Mapping):
+                mapped_resources.append(self._forge.map(resource_dict, type_))
+            elif type_ in mappings:
+                # type_ is the entity here
+                mapping_class : Mapping = import_class(mappings[type_][0], "mappings")
+                mapping = self._model.mapping(type_, self._model.source, mapping_class)
+                mapped_resources.append(self._forge.map(resource_dict, mapping))
+            else:
+                mapped_resources.append(resource)
+        return mapped_resources
+
+    def types(self):
+        # TODO: add other datatypes used, for instance, inside the mappings
+        return list(self.mappings().keys())
+
+    @abstractmethod
+    def search(self, resolvers, *filters, **params) -> Resource:
```

Comment:
These two methods are mainly to be able to do mappings directly after searching resources from the databases.